### PR TITLE
Multi input handling

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will update the multi-input handling capabilities of HELICS and do some code refactoring around the related code

### Proposed changes

- refactor multi-input handing code in federateState and Core.
- some refactoring in the valueFederate handlers.  
- fix associated clang-tidy issues

### CODE REVIEW

**Comment by phlptp** [comment]
@nightlark @kdheepak  I am trying to work through some things for handling multiple inputs to a single input.    To allow some customization of how that is handled we need a little more flexibility in setting interface properties.    There are two ways I see of doing this

right now 
`
void helicsInputSetOption(helics_input inp, int option, helics_bool value, helics_error* err)
`

proposed

`void helicsInputSetOption(helics_input inp, int option, int value, helics_error* err)`  

The int option may be an enumeration from @kdheepak current pr
This is not a signature change in C but it does change the semantics of allowing more options than just a bool.   There are other changes on the C++ side of actual bool to int.  Bool would then be `0` or `not 0`

The other option is to add a new function to allow more flexible property settings I would prefer to not add another set of 4 methods if at all possible but it appears we do need some more flexibility in this case. 

This is needed to allow multi-input processing specification for different vector collapse arguments for different inputs.

**Comment by phlptp** [comment]
This is the first of 3 PR's on this topic.  The second will add tests and functionality for the C and C++98 API.  And the third will add documentation.

File: src/helics/application_api/Federate.hpp [hunk 1]
```diff
@@ -535,13 +535,15 @@ class HELICS_CXX_EXPORT Federate {
     /** set an interface option
     @param handle the handle of an interface to modify
     @param option the option field to modify
-    @param option_value the value to set(default to 1 or true most options are boolean but some allow for other values
+    @param option_value the value to set(default to 1 or true most options are boolean but some
```
Comment (nightlark) [inline src/helics/application_api/Federate.hpp:538]:
```suggestion
    @param option_value the value to set (defaults to 1 or true), most options are boolean but some
```

File: src/helics/core/InputInfo.hpp [hunk 2]
```diff
@@ -51,36 +64,37 @@ class NamedInputInfo {
...
+        current_data_time;  //!< the most recent published data times
+    std::vector<std::shared_ptr<const data_block>>
+        current_data;  //!< the most recent published data
     std::vector<global_handle> input_sources;  //!< the sources of the input signals
-    std::vector<Time> deactivated;
-    std::vector<std::tuple<std::string, std::string, std::string>>
-        source_info;  //!< the name,type,units of the sources
+    std::vector<Time> deactivated;  //!< indicator that the source has been deactivated
+    std::vector<sourceInformation> source_info;  //!< the name,type,units of the sources
+    std::vector<int32_t> priority_sources;  //!< the list or priority inputs;
```
Comment (nightlark) [inline src/helics/core/InputInfo.hpp:87]:
```suggestion
    std::vector<int32_t> priority_sources;  //!< the list of priority inputs;
```

File: src/helics/application_api/Federate.cpp [hunk 3]
```diff
@@ -1243,7 +1243,7 @@ void Federate::setInterfaceOption(interface_handle handle, int32_t option, bool
 }
 
 /** get the current value for an interface option*/
-bool Federate::getInterfaceOption(interface_handle handle, int32_t option)
+int Federate::getInterfaceOption(interface_handle handle, int32_t option)
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/application_api/Federate.cpp:1246]: probably

File: src/helics/application_api/Inputs.cpp [hunk 4]
```diff
@@ -115,26 +116,417 @@ void Input::handleCallback(Time time)
...
+                res.push_back('"');
+                res.push_back(',');
+            }
+            if (res.back() == ',') {
+                res.back() = ']';
+            } else {
+                res.push_back(']');
+            }
+            return res;
+        }
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/application_api/Inputs.cpp:292]: I will look into just using the jsoncpp function to create this, that may be simpler than trying to deal with all the appropriate edge cases

File: src/helics/helics_enums.h [hunk 5]
```diff
@@ -225,7 +227,15 @@ typedef enum {
     /** specify that an interface will only update if the value has actually changed*/
     helics_handle_option_only_update_on_change = 8,
     /** specify that an interface does not participate in determining time interrupts*/
-    helics_handle_option_ignore_interrupts = 475
+    helics_handle_option_ignore_interrupts = 475,
+    /** specify the multi-input processing method for inputs*/
+    helics_handle_option_multi_input_handling_method = 507,
+    /** specify the input with highest priority*/
+    helics_handle_option_input_priority_location = 510,
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/helics_enums.h:234]: Still playing with how this all might work while minimizing the number of new API calls.  The idea was to have this flag set the index of the highest priority,  Subsequent calls override it but establish a priority ranking if desired.

File: src/helics/helics_enums.h [hunk 6]
```diff
@@ -225,7 +227,15 @@ typedef enum {
...
+    /** specify the input with priority*/
+    helics_handle_option_clear_priority_list = 512,
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/helics_enums.h:236]: I will fix it

**Comment by phlptp** [comment]
@nightlark any more changes you would recommend.  There will be a second PR that deals with the C interface and any more changes required there.

**Pull request merged**